File: cpp/TopoSort.cpp
#include <vector>
#include <queue>

class TopologicalSort
{
    std::vector<std::vector<int>> G;
    std::vector<int> indeg;

public:
    TopologicalSort(int n) : G(n), indeg(n, 0) {}

    void add_edge(int s, int t)
    {
        G[s].emplace_back(t);
        indeg[t]++;
    }
    std::vector<int> build()
    {
        int n = (int)G.size();
        std::queue<int> q;
        // use prority to get lexicographical order
        std::vector<bool> used(n, false);

        auto push = [&](int u)
        {
            if (used[u])
            {
                return;
            }
            q.emplace(u);
            used[u] = true;
        };

        for (int i = 0; i < n; i++)
        {
            if (indeg[i] == 0)
            {
                push(i);
            }
        }

        std::vector<int> ps;
        while (not q.empty())
        {
            int u = q.front();
            q.pop();
            ps.emplace_back(u);
            for (int v : G[u])
            {
                if (--indeg[v] == 0)
                {
                    push(v);
                }
            }
        }
        if (n != (int)ps.size())
        {
            return {};
        }
        return ps;
    }
};

File: python/TopoSort.py
from collections import deque
import heapq


class TopologicalSort:
    def __init__(self, n: int) -> None:
        self.G = [[] for _ in range(n)]
        self.indeg = [0]*n

    def add_edge(self, s: int, t: int):
        self.G[s].append(t)
        self.indeg[t] += 1

    def build(self):
        n = len(self.G)
        q = deque()
        # q=[] heapq
        used = [False]*n

        def push(u):
            if used[u]:
                return
            q.append(u)
            # heapq.heappush(q, u)
            used[u] = True

        for i in range(n):
            if self.indeg[i] == 0:
                push(i)

        ps = []
        while q:
            u = q.popleft()
            # u = heapq.heappop(q)
            ps.append(u)
            for v in self.G[u]:
                self.indeg[v] -= 1
                if self.indeg[v] == 0:
                    push(v)
        if n != len(ps):
            return []
        return ps
